aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/AWSCognitoAuthPlugin.java

```diff
@@ -0,0 +1,430 @@
+                }
+
+                @Override
+                public void onError(Exception error) {
+                    onException.accept(
+                        new AmplifyException(
+                                "Sign in with Facebook failed",
+                                error,
+                                "See attached exception for more details"
+                        )
+                    );
+                }
+            }
+        );
+    }
+
+    // The result of a success callback is an object of type AWSCognitoAuthSession so that when the result
+    // is cast to that type, the following Cognito specific fields can be retrieved:
+    //   - AWSCredentials
+    //   - Cognito Identity ID
+    //   - Cognito Access Token
+    //   - Cognito ID Token
+    //   - Cognito Refresh Token
+    @Override
+    public void currentAuthState(
+            @NonNull Consumer<AuthSession> onSuccess,
+            @NonNull Consumer<AuthException> onError
+    ) {
+        try {
+            AWSMobileClient.getInstance().currentUserState(new Callback<UserStateDetails>() {
+                @Override
+                public void onResult(UserStateDetails result) {
+                    UserState state = result.getUserState();
+
+                    if (state.equals(UserState.SIGNED_OUT) ||
+                            state.equals(UserState.SIGNED_OUT_FEDERATED_TOKENS_INVALID) ||
+                            state.equals(UserState.SIGNED_OUT_USER_POOLS_TOKENS_INVALID)
+                    ) {
+                        onSuccess.accept(new AWSCognitoAuthSession(AuthUserState.SIGNED_OUT));
+                    } else {
+                        final AWSCognitoAuthSession session;
+                        final CountDownLatch latch = new CountDownLatch(2);
+
+                        if (state.equals(UserState.SIGNED_IN) || state.equals(UserState.GUEST)) {
+                            session = new AWSCognitoAuthSession(
+                                    state.equals(UserState.SIGNED_IN)
+                                            ? AuthUserState.SIGNED_IN
+                                            : AuthUserState.GUEST,
+                                    null,
+                                    AWSMobileClient.getInstance().getIdentityId(),
+                                    null,
+                                    null,
+                                    null
+                            );
+
+                            AWSMobileClient.getInstance().getTokens(new Callback<Tokens>() {
+                                @Override
+                                public void onResult(Tokens result) {
+                                    session.setAccessToken(result.getAccessToken().getTokenString());
+                                    session.setIdToken(result.getIdToken().getTokenString());
+                                    session.setRefreshToken(result.getRefreshToken().getTokenString());
```

Comment:
Maybe you want a utility method where you pass in a `Tokens`, and get back a fully populated `AWSCognitoAuthSession ` object. I guess, `AWSCognitoAuthSessionFactory` would be a good name for it?

```
...
    public static AWSCognitoAuthSession fromTokens(Tokens tokens) {
       AWSCognitoAuthSession session = new AWSCognitoAuthSession();
       ... set fields
       return session;
    }
```

This would be economical from a software craft standpoint if you will use the utility ~3+ times, probably?

Reply:
The problem with this is that I need to populate some of the AWSCognitoAuthSession object before this callback and also need to populate some of it potentially after this (depending on what order the AWSMobileClient calls come back)